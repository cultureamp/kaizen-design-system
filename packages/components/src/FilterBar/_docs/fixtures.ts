/* eslint-disable @typescript-eslint/explicit-function-return-type */
import { Factory, faker } from "@cultureamp/frontend-testing"
import { FilterOption } from "../types"

const userFactory = Factory.Sync.makeFactory<FilterOption>({
  value: Factory.each(i => faker.datatype.uuid()),
  label: Factory.each(
    () => `${faker.name.firstName()} ${faker.name.lastName()}`
  ),
})

const departmentFactory = Factory.Sync.makeFactory<FilterOption>({
  value: Factory.each(i => faker.datatype.uuid()),
  label: Factory.each(() => faker.commerce.department()),
})

const ascSort = (a: FilterOption, b: FilterOption) =>
  a.label.toLocaleLowerCase() > b.label.toLocaleLowerCase() ? 1 : -1

export const users = userFactory.buildList(30).sort(ascSort)
export const departments = departmentFactory.buildList(8).sort(ascSort)
