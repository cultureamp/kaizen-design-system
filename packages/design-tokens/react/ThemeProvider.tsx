import React from "react"
import { ThemeManager, defaultTheme, Theme } from "../"

export const ThemeContext = React.createContext<Theme>(defaultTheme)

/**
 * Wrap your application in this provider using a ThemeManager, to synchronise it with a react context.
 * This allows child react elements to more easily use theme variables, using the {@link useTheme} hook.
 */
export const ThemeProvider = ({
  themeManager,
  ...props
}: {
  themeManager: ThemeManager
  children: React.ReactNode
}) => {
  const [theme, setTheme] = React.useState<Theme>(
    themeManager.getCurrentTheme()
  )
  React.useEffect(() => {
    let cancelled = false
    const listener = (newTheme: Theme) => {
      if (!cancelled) setTheme(newTheme)
    }
    themeManager.addThemeChangeListener(listener)
    return () => {
      cancelled = true
      themeManager.removeThemeChangeListener(listener)
    }
  }, [])

  return (
    <ThemeContext.Provider value={theme}>
      {props.children}
    </ThemeContext.Provider>
  )
}

export const useTheme = () => React.useContext(ThemeContext)
