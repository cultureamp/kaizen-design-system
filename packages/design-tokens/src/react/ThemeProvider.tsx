import React, { useState } from "react"
import { ThemeManager, defaultTheme, Theme } from ".."

export const ThemeContext = React.createContext<Theme>(defaultTheme)

/**
 * Wrap your application in this provider using a ThemeManager, to synchronise it with a react context.
 * This allows child react elements to more easily use theme variables, using the {@link useTheme} hook.
 */
/**
 * @deprecated ThemeProvider is deprecated. Please use KaizenProvider from "@kaizen/components" instead.
 */
export const ThemeProvider = ({
  themeManager,
  ...props
}: {
  themeManager?: ThemeManager
  children: React.ReactNode
}): JSX.Element => {
  const [themeManagerInstance] = useState(
    () => themeManager || new ThemeManager(defaultTheme)
  )

  const [theme, setTheme] = React.useState<Theme>(
    themeManagerInstance.getCurrentTheme()
  )
  React.useEffect(() => {
    let cancelled = false
    const listener = (newTheme: Theme): void => {
      if (!cancelled) setTheme(newTheme)
    }
    themeManagerInstance.addThemeChangeListener(listener)
    return () => {
      cancelled = true
      themeManagerInstance.removeThemeChangeListener(listener)
    }
  }, [])

  return (
    <>
      <ThemeContext.Provider value={theme}>
        {props.children}
      </ThemeContext.Provider>
      <link
        rel="preload"
        href="https://d1e7r7b0lb8p4d.cloudfront.net/fonts/inter/inter-bold.woff2"
        as="font"
        type="font/woff2"
        crossOrigin="anonymous"
      />
      <link
        rel="preload"
        href="https://d1e7r7b0lb8p4d.cloudfront.net/fonts/inter/inter-demi-bold.woff2"
        as="font"
        type="font/woff2"
        crossOrigin="anonymous"
      />
      <link
        rel="preload"
        href="https://d1e7r7b0lb8p4d.cloudfront.net/fonts/inter/inter-medium.woff2"
        as="font"
        type="font/woff2"
        crossOrigin="anonymous"
      />
      <link
        rel="preload"
        href="https://d1e7r7b0lb8p4d.cloudfront.net/fonts/inter/inter-regular.woff2"
        as="font"
        type="font/woff2"
        crossOrigin="anonymous"
      />
      <link
        rel="preload"
        href="https://d1e7r7b0lb8p4d.cloudfront.net/fonts/tiempos/tiempos-headline-bold.woff2"
        as="font"
        type="font/woff2"
        crossOrigin="anonymous"
      />
    </>
  )
}

export const useTheme = (): Theme => React.useContext(ThemeContext)
