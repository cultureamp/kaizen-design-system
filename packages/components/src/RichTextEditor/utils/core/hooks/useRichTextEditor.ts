import { useCallback, useEffect, useRef, useState } from "react"
import { EditorState } from "prosemirror-state"
import { createRichTextEditor } from "../create"
import { CommandOrTransaction } from "../types"

type Options = {
  editable: boolean
}

type SetEditableStatus = (status: boolean) => void

type UseRichTextEditorReturnValue = [
  any,
  EditorState,
  (commandOrTransaction: CommandOrTransaction) => void,
  SetEditableStatus,
]

/**
 * useRichTextEditor
 * React hook to initialize a ProseMirror editor, handle binding it to the DOM,
 * and updating the state within React’s lifecycle.
 *
 * @param {initialEditorState} ProseMirror state
 * @returns {Array}
 */
export function useRichTextEditor(
  initialEditorState: EditorState,
  /*
   * Pass in HTML attributes into the parent RTE node
   */
  attributes?: {
    [name: string]: string
  },
  options?: Options
): UseRichTextEditorReturnValue {
  options = {
    editable: true,
    ...options,
  }
  const [editorState, setEditorState] =
    useState<EditorState>(initialEditorState)
  // Refs to hold the methods returned from ProseMirror’s initialization
  const destroyEditorRef = useRef<() => void>()
  const dispatchTransactionRef = useRef<
    (commandOrTransaction: CommandOrTransaction) => void
  >(() => undefined)

  // Construct a consistent reference to call the dispatchTransactionRef without
  // forcing the consumer to unwind it
  const dispatchTransaction = useCallback(
    (commandOrTransaction: CommandOrTransaction) => {
      dispatchTransactionRef.current(commandOrTransaction)
    },
    [dispatchTransactionRef]
  )

  // Hold editableStatus as a ref so we can toggle its status
  const editableStatusRef = useRef<boolean>(options.editable)
  const setEditableStatus = useCallback<SetEditableStatus>(
    status => {
      editableStatusRef.current = status
      // Trigger an update within ProseMirror by issuing a noop transaction
      dispatchTransaction((state, dispatch) => {
        if (!dispatch) return false
        dispatch(state.tr)
        return true
      })
    },
    [editableStatusRef]
  )

  const editorRef = useCallback(
    (node: HTMLElement) => {
      if (node !== null) {
        const instance = createRichTextEditor({
          node,
          initialEditorState: editorState,
          onChange: setEditorState,
          isEditable: () => editableStatusRef.current,
          attributes,
        })
        destroyEditorRef.current = instance.destroy
        dispatchTransactionRef.current = instance.dispatchTransaction
      }
    },

    // Including editorState in the dependencies here will cause an endless
    // loop as the initialization changes its value
    [setEditorState, editableStatusRef]
  )

  // Tear down ProseMirror when the consuming component is unmounted
  useEffect(
    () => () => {
      if (destroyEditorRef.current) {
        destroyEditorRef.current()
      }
    },
    [destroyEditorRef]
  )

  return [editorRef, editorState, dispatchTransaction, setEditableStatus]
}
