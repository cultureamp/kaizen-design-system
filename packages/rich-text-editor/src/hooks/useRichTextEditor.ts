import { EditorState } from "prosemirror-state"
import { useCallback, useEffect, useRef, useState } from "react"
// import { MaybeCommand } from "../types.d"
import { createRichTextEditor } from "../create"

export function useRichTextEditor(
  initialEditorState: EditorState
): [any, EditorState, (maybeCommand: any) => void] {
  const [editorState, setEditorState] = useState<EditorState>(
    initialEditorState
  )
  const destroyEditorRef = useRef<() => void>()
  const dispatchTransactionRef = useRef<(maybeCommand: any) => void>(
    () => undefined
  )

  const editorRef = useCallback(
    (node: HTMLElement) => {
      if (node !== null) {
        const instance = createRichTextEditor({
          node,
          initialEditorState: editorState,
          onChange: setEditorState,
        })
        // We need to set these lazily to ensure useState retains the functions
        // as values (otherwise they get evaluated immediately)
        // https://reactjs.org/docs/hooks-reference.html#lazy-initialization
        destroyEditorRef.current = instance.destroy
        dispatchTransactionRef.current = instance.dispatchTransaction
      }
    },

    // Including editorState in the dependencies here will cause explosions
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [setEditorState]
  )

  // Destroy when the consuming component is unmounted
  useEffect(
    () => () => {
      if (destroyEditorRef.current) {
        destroyEditorRef.current()
      }
    },
    [destroyEditorRef]
  )

  return [editorRef, editorState, dispatchTransactionRef.current]
}
