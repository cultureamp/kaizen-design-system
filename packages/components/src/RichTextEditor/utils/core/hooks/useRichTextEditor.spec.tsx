import React, { useEffect } from "react"
import { act, configure, render, screen } from "@testing-library/react"
import { EditorState, Transaction } from "prosemirror-state"
import { testEditorState } from "../fixtures/testState"
import { useRichTextEditor } from "./useRichTextEditor"

configure({ testIdAttribute: "data-automation-id" })

function Scenario({
  onChange = (_: EditorState) => undefined,
  editable = true,
}: {
  onChange?: (editorState: EditorState) => void
  editable?: boolean
}): JSX.Element {
  const command = (
    state: EditorState,
    dispatch?: (tx: Transaction) => void
  ) => {
    // Insert text at the current selection point, which is the start because
    // we donâ€™t have a selection yet.
    if (!dispatch) return false
    dispatch(state.tr.insertText("Prepended content. "))
    return true
  }

  const [ref, editorState, dispatchTransaction, setEditableStatus] =
    useRichTextEditor(
      testEditorState,
      { "aria-labelledby": "label-ref-id", "data-test-id": "12345678" },
      { editable }
    )

  useEffect(() => {
    // Propagate changes to the editorState
    onChange(editorState)
  }, [editorState])

  return (
    <div>
      <button type="button" onClick={() => dispatchTransaction(command)}>
        Prepend button
      </button>
      <button type="button" onClick={() => setEditableStatus(false)}>
        Editable: false
      </button>
      <button type="button" onClick={() => setEditableStatus(true)}>
        Editable: true
      </button>
      <div data-automation-id="editor" ref={ref} />
    </div>
  )
}

describe("useRichTextEditor", () => {
  it("binds the editor to the DOM", async () => {
    render(<Scenario />)

    await screen.findByText("Example content")
  })

  it("updates the DOM when commands are dispatched", async () => {
    render(<Scenario />)

    await act(async () => {
      await screen.findByText("Example content")
    })

    const button = await screen.findByText("Prepend button")
    button.click()

    await screen.findByText("Prepended content. Example content")
  })

  it("updates the editorState when commands are dispatched", async () => {
    const onChange = jest.fn()
    render(<Scenario onChange={onChange} />)

    await act(async () => {
      await screen.findByText("Example content")
    })

    await act(async () => {
      const button = await screen.findByText("Prepend button")
      button.click()
    })

    const updatedEditorState = onChange.mock.calls[1][0] as EditorState
    expect(updatedEditorState.toJSON()).toMatchSnapshot()
  })

  it("defaults to editable", async () => {
    render(<Scenario />)

    const editor = await screen.findByTestId("editor")
    expect(editor.children[0]?.getAttribute("contenteditable")).toEqual("true")
  })

  it("respects initial editable status", async () => {
    render(<Scenario editable={false} />)

    const editor = await screen.findByTestId("editor")
    expect(editor.children[0]?.getAttribute("contenteditable")).toEqual("false")
  })

  it("updates editable status", async () => {
    render(<Scenario />)

    const editor = await screen.findByTestId("editor")
    expect(editor.children[0]?.getAttribute("contenteditable")).toEqual("true")

    const disableEditButton = await screen.findByText("Editable: false")
    disableEditButton.click()

    expect(editor.children[0]?.getAttribute("contenteditable")).toBe("false")

    const enableEditButton = await screen.findByText("Editable: true")
    enableEditButton.click()

    expect(editor.children[0]?.getAttribute("contenteditable")).toBe("true")
  })
})
