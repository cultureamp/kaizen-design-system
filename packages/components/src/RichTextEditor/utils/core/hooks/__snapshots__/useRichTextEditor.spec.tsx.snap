// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useRichTextEditor updates the editorState when commands are dispatched 1`] = `
Object {
  "doc": Object {
    "content": Array [
      Object {
        "content": Array [
          Object {
            "text": "Prepended content. Example content",
            "type": "text",
          },
        ],
        "type": "paragraph",
      },
    ],
    "type": "doc",
  },
  "selection": Object {
    "anchor": 20,
    "head": 20,
    "type": "text",
  },
}
`;
